### DISCUSSION TITLE
Changing of formulas in v0.25.8 causing trouble with old code

### DISCUSSION THREAD

**Discussion by helangen** [CamDavidsonPilon/lifelines#1237] [Q&A] [answered]
MacOS BigSur v11.2.1
Python v3.8.3
Lifelines v0.25.9

Hi,
I had a univariate and multivariate Cox-regression set up with formulas. After updating this code does not work and throws an error: `Unable to evaluate factor `C(Sex, Treatment(0))`. [NameError: name 'Treatment' is not defined]`
I assume this is the change from Patsy to Formulaic.

My current code is:
`cph_uni_sex.fit(pd_cox_regression, duration_col='OS (days)', event_col='Status (0/1)', formula = "C(Sex, Treatment(0))")`

I know a formula is not really neccessary for this problem, but I have further regressions with groups and I need to specify. which group to compare against. My problem now is how I get this to work with the new Formulaic language. I can't find any examples in the documentation, and trying to read the documentation for Formulaic got me even more confused (https://cran.r-project.org/web/packages/formulaic/vignettes/Introduction-to-formulaic.html). I can always downgrade to a previous version, but I wanted to get it working again with the new version.

Any help is appreciated.

**Comment by CamDavidsonPilon** [comment, marked as answer]
Hi @helangen! First off, apologies that this change broke your code. I wanted the transition to a new parsing library to be smooth for the end user - but here we are. Let's see what we can do. 

The parsing library I'm using is Python's [formulaic](https://matthewwardrop.github.io/formulaic/basic/grammar/) - not R's. Formulaic implements the same language as Patsy (the previous parser). However it's not as feature complete as Patsy. For example, it only offers [partial (see note 6)](https://matthewwardrop.github.io/formulaic/basic/grammar/#fn:6) support for custom contrast matrices (as you are doing here). 

So what can we do? 

1. Naive: drop the `Treatment(0)` part of the formula. Inference will still proceed, but it's obvious that you want that, so dropping it seems silly.
2. Move encoding outside lifelines, and you can use Patsy again. Something like:
  ```
  from patsy import dmatrix
  X = dmatrix("Q('OS (days)') +  Q('Status (0/1)') +  C(Sex, Treatment(0))", pd_cox_regression, return_type="dataframe")
  X = X.drop("Intercept", axis=1)
  cph_uni_sex.fit(X, duration_col="Q('OS (days)')", event_col="Q('Status (0/1))")
  ```
3. Downgrade lifelines to <del>0.24.8</del> 0.25.7

Each option has a tradeoff, but I think 3. is the least headache. Is there something in the previous two versions that you really want? I can help bridge that gap.

**Reply by helangen** [reply]
Hi. Thanks for that. I think the easiest option is to downgrade which is not a huge problem for this project. I guess downgrading to 0.25.7 would suffice or what since it seems like the changes was implemented in v0.25.8? v0.24.8 seems like a huge downgrade. I could keep an eye on formulaic for changes and then upgrade at a later stage when it is more mature.

**Reply by CamDavidsonPilon** [reply]
whoops, I did make a mistake in the version to downgrade to: `0.25.7` is correct.
